projects/components/src/datagrid/datagrid.component.ts

```diff
@@ -627,28 +635,35 @@ export class DatagridComponent<R> implements OnInit, AfterViewInit {
     private updateSelectedItems(): void {
         if (this._selectionType === GridSelectionType.Single && this.datagrid.selection.currentSingle) {
             // Tries to find the currently selected item. If it isn't found, clears the selection.
+            const current = this.datagrid.selection.currentSingle;
             const found = this.items.find(
                 (item, itemIndex) =>
-                    this.trackBy(itemIndex, item) ===
-                    this.trackBy(
-                        this.items.indexOf(this.datagrid.selection.currentSingle),
-                        this.datagrid.selection.currentSingle
-                    )
+                    this.trackBy(itemIndex, item) === this.trackBy(this.items.indexOf(current), current)
             );
             if (!found) {
-                this.datagrid.selection.currentSingle = undefined;
+                this.datagrid.selection.clearSelection();
+            } else {
+                this.datagrid.selection.setSelected(found, true);
             }
         } else if (this._selectionType === GridSelectionType.Multi) {
             // Tries to find the currently selected items. If an item isn't found, clears the selection for that item.
             if (this.datagrid.selection.current) {
-                this.datagrid.selection.current = this.datagrid.selection.current.filter((selected, selectedIndex) => {
-                    const found = this.items.find(
-                        (item, itemIndex) => this.trackBy(itemIndex, item) === this.trackBy(selectedIndex, selected)
-                    );
-                    return found;
-                });
+                const current = [...this.datagrid.selection.current];
+                this.datagrid.selection.clearSelection();
+                const nextSelection = current
```

Comment:
Add a comment explaining that we are mapping from the old records to the new records since they will typically new references.

It may be worth an appropriately named private method

```ts

/**
 * Given an existing selection that is made up of records that are about to become stale, and new records to be loaded into the grid, returns the selection mapped to records from the newly added records, excluding any records that may not be present in the new selection because they are not present in the new data.
 */
mapSelectedRecords(currentSelection: T[], newRecords: T[]): T[] {

}
```